src/iso_alloc_sanity.c

```diff
@@ -142,19 +142,37 @@ struct uffdio_api _uffd_api;
 int64_t _uf_fd;
 
 INTERNAL_HIDDEN void _iso_alloc_setup_userfaultfd(void) {
-    _uf_fd = syscall(__NR_userfaultfd, O_CLOEXEC | O_NONBLOCK);
+    int flags = O_CLOEXEC | O_NONBLOCK;
+
+#ifdef UFFD_USER_MODE_ONLY
+    /* Below, the syscall fails not necessarily because lack of kernel support.
+     * By default, the `unprivileged_userfaultfd` flag is disabled.
+     * From Linux 5.11 however, we can handle pages registered from the user-space. */
+    flags |= UFFD_USER_MODE_ONLY;
+#endif
+    _uf_fd = syscall(__NR_userfaultfd, flags);
 
     if(_uf_fd == ERR) {
-        LOG_AND_ABORT("This kernel does not support userfaultfd");
+        LOG_AND_ABORT("This kernel does not support userfaultfd or is disallowed in the user-space");
     }
 
     _uffd_api.api = UFFD_API;
     _uffd_api.features = 0;
 
+#if THREAD_SUPPORT
+    _uffd_api.features |= UFFD_FEATURE_THREAD_ID;
+#endif
+
     if(ioctl(_uf_fd, UFFDIO_API, &_uffd_api) == ERR) {
         LOG_AND_ABORT("Failed to setup userfaultfd with ioctl");
     }
 
+#if THREAD_SUPPORT
+    if(!(_uffd_api.features & UFFD_FEATURE_THREAD_ID)) {
```

Comment:
Does the `ioctl` call reset this if it cannot be used?

Reply:
[seems to be the case](https://github.com/torvalds/linux/blob/ed30a4a51bb196781c8058073ea720133a65596f/fs/userfaultfd.c#L2091).

Reply:
Should we also check `errno` for `-EFAULT` here?

Reply:
not sure we really need to, return value is checked for ioctl already.